src/core/lib/slice/slice_hash_table.c

```diff
@@ -29,78 +29,87 @@
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 
-#include "src/core/lib/transport/mdstr_hash_table.h"
+#include "src/core/lib/slice/slice_hash_table.h"
 
 #include <stdbool.h>
 #include <string.h>
 
 #include <grpc/support/alloc.h>
 #include <grpc/support/log.h>
 
+#include "src/core/lib/slice/slice_internal.h"
 #include "src/core/lib/transport/metadata.h"
 
-struct grpc_mdstr_hash_table {
+struct grpc_slice_hash_table {
   gpr_refcount refs;
   size_t size;
-  grpc_mdstr_hash_table_entry* entries;
+  grpc_slice_hash_table_entry* entries;
 };
 
+static bool is_empty(grpc_slice_hash_table_entry* entry) {
+  return entry->vtable == NULL;
+}
+
 // Helper function for insert and get operations that performs quadratic
 // probing (https://en.wikipedia.org/wiki/Quadratic_probing).
-static size_t grpc_mdstr_hash_table_find_index(
-    const grpc_mdstr_hash_table* table, const grpc_mdstr* key,
-    bool find_empty) {
+static size_t grpc_slice_hash_table_find_index(
+    const grpc_slice_hash_table* table, const grpc_slice key, bool find_empty) {
+  size_t hash = grpc_slice_hash(key);
   for (size_t i = 0; i < table->size; ++i) {
-    const size_t idx = (key->hash + i * i) % table->size;
-    if (table->entries[idx].key == NULL) return find_empty ? idx : table->size;
-    if (table->entries[idx].key == key) return idx;
+    const size_t idx = (hash + i * i) % table->size;
+    if (is_empty(&table->entries[idx])) {
+      return find_empty ? idx : table->size;
+    }
+    if (grpc_slice_cmp(table->entries[idx].key, key) == 0) {
```

Comment:
Updated to the new grpc_slice_eq: it's also constant time (not quite a pointer equality, but not far off) if the two slices are interned, and falls back to string comparison only when needed.